### ISSUE TITLE
Why is Segmentation Training in YOLOv8 not good?

### ISSUE THREAD

**Issue by Kader95Sh** [ultralytics/ultralytics#2090]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hallo,

I trained YOLOv8x-seg and YOLOv7x-seg on the same dataset with exact same labels and images for Training, validation and test. 

The result for YOLOv8 is not good at all, even after 200 Epoch. YOLOv7 gave much better Results after 2 epochs only.

Hier is the result of YOLOv8 after 200 epochs
![yolov8](https://user-images.githubusercontent.com/86665915/232603600-80de6e03-0aab-461a-9d09-0525d50ac3be.png)

Hier for YOLOv7 after2 epochs
![yolov7](https://user-images.githubusercontent.com/86665915/232603917-a32e2539-eece-4ada-b9a0-d5cd02e3194b.png)

i would like to ask you, why is that happening? In both of them I am using the default training without chainging anything, just image size to 1408.

Is there any way to Train YOLOv8x-seg to be better?

Thank you so much.

**Comment by glenn-jocher** [comment]
@Kader95Sh hi there,

Thank you for reaching out. I suggest checking out the documentation at https://docs.ultralytics.com/. There you will find more information about how to customize your training such as adjusting the learning rate and other hyperparameters. Additionally, it is possible that YOLOv8 may perform differently on your specific dataset compared to YOLOv7. You may need to experiment with different settings to achieve higher performance with YOLOv8. I hope this helps!

**Comment by Laughing-q** [comment]
@Kader95Sh hi, if training results are not as expected, the first step for me is to check the `train_batch*.jpg` to see if the labels are all good. Then check some default hyps which might damage the training results of your dataset.
BTW I'm not sure which repo of yolov7-seg you are using but as far as I know most of the segmentation implementations of yolo series are based on yolov5-seg which is inspired by yolact, so is yolov8-seg. Hence empirically I think there should not be such a gap of the results. 
There probably are some hyps we have that are not good for your case, this is dataset-specific I'm not able to give much ideas.
Still there're some general tips you can try:
- try higher image resolution during training.
- not sure if you're using a pretrained weights as initialized weights, but results would be better if you use one.
- add `retina_masks=True` when you predict segmentation, it'll produce masks with higher quality than default.
-  more tips please refer to our docs page [https://docs.ultralytics.com/yolov5/tips_for_best_training_results/](https://docs.ultralytics.com/yolov5/tips_for_best_training_results/) that we migrated from yolov5 and it's still effective in yolov8.

Good luck! :)

**Comment by Kader95Sh** [comment]
Thank you for your answers.

I tried already those solutions.

How do I run YOLOv7:
srun python3 segment/train.py --cache disk --data custom22.yaml --workers 4 --batch 4 --weights 'yolov7-seg.pt' --cfg yolov7-seg.yaml --epochs 200 --device 0 --name KITTI --img 1408 --hyp data/hyps/hyp.scratch-high.yaml

How do I run YOLOv8:
srun yolo task=segment mode=train cache='disk' model=yolov8x-seg.pt device=\'0,1\' data=./custom22.yaml epochs=200 batch=16 imgsz=1408 save_period=50 workers=8 project=kitti360V2 name=BigSegX patience=50

I tried also to use the same Hyperparameters from YOLOv7 in YOLOv8.

I am getting same results.

Did I miss any of your advices?

I also added some new Labels 'sky' and 'road' and this is the train_batch from YOLOv8.
![train_batch0](https://user-images.githubusercontent.com/86665915/233257836-f22f086b-97d9-4a66-84ec-dfb8634c0908.jpg)

Best regards

**Comment by glenn-jocher** [comment]
@Kader95Sh thank you for providing more information. It's good to know that you have tried the suggestions already. It's possible that YOLOv8 may perform differently on your specific dataset compared to YOLOv7 because they have different network architectures and hyperparameters. It's always recommended to experiment with different settings to achieve the best performance for your specific use case. Another suggestion is to try using a different pre-trained weight file as your initial weights for training YOLOv8, this may help improve your results. Additionally, you can also try increasing the batch size or changing the learning rate to see if it helps. Lastly, you may want to consider other segmentation models that might perform better on your dataset. I hope this helps, and good luck with your experimentation!

**Comment by Laughing-q** [comment]
@Kader95Sh is this train_batch what you expected? the train_batch image looks weird to me, there are many serrations here🤔

**Comment by Kader95Sh** [comment]
No it is not as i expected. Thats right @Laughing-q , there is many serrations and I dont know why is that happening. The prediction looks like the train_batch.

**Comment by Laughing-q** [comment]
@Kader95Sh does the labels looks like the same(also many serrations) in your v7 training?

**Comment by glenn-jocher** [comment]
@Kader95Sh it's good to know that the issue is not from the labels. The images you provided are from the KITTI-360 dataset, and it looks like there are some compression artifacts in the images that may cause issues during training. One possible solution could be to resize the images and apply some denoising techniques to help clean up these artifacts before training. Additionally, you may want to try different augmentation techniques during training, such as adding more noise or distortion, to help your model generalize better. Another option could be using an alternate dataset or pre-training the model with a large-scale image classification dataset such as ImageNet, which might help improve the model's generalization ability. I hope this helps you, and good luck with your experimentation!

**Comment by Laughing-q** [comment]
@Kader95Sh looks like your labels are incorrect. The segment labels you provided should be boundary coordinates of mask area but it seems your labels are the whole coordinates inside the mask area.
it should be some thing like this(`000000000025.txt` from our example dataset coco128):
![1](https://user-images.githubusercontent.com/61612323/233526036-47eba7e0-aabe-4306-9013-43d56ad2747a.jpg)
but your label looks like:
![1](https://user-images.githubusercontent.com/61612323/233526098-4f3d9f75-90b8-41df-8be1-5e6160bf9e7c.jpg)

here's the script I visualize the labels:
```python
import cv2
import numpy as np

image = cv2.imread("./00f0000000250.jpg")
with open("./00f0000000250.txt") as f:
    segment = [np.array(x.split(), dtype=np.float32)[1:].reshape(-1, 2) for x in f.read().strip().splitlines() if len(x)]
h, w = image.shape[:2]
for s in segment:
    s[:, 0] *= w
    s[:, 1] *= h
cv2.drawContours(image, [s.astype(np.int32) for s in segment], -1, (0, 0, 255), thickness=2)
cv2.imwrite("1.jpg", image)
```

**Comment by Kader95Sh** [comment]
@Laughing-q Thank you for you answers and help.
I tried this kind of labels at the begin and the results was worse. 
train_batch0:
![train_batch0](https://user-images.githubusercontent.com/86665915/233615480-7225f645-24df-47a1-abcd-55b7a01379d4.jpg)

results:
![image](https://user-images.githubusercontent.com/86665915/233615653-10973c6a-ec5a-4bbc-b15f-d623a01f28ed.png)

But after I used your script to visualize the labels, it would looks like --> 
![image](https://user-images.githubusercontent.com/86665915/233615979-08a89761-a7f9-4684-a808-97b90dd2de2c.png)

So if YOLOv8 is using same way to train the data, I think it will be the problem that i am facing.
Should labels have kind of sort to look like your example dataset coco128?

If i visualize the labels by just chang the pixle color it will look like -->
![image](https://user-images.githubusercontent.com/86665915/233616681-3851f878-b2a9-4c21-9253-a6c8eb4480f1.png)

best regards

**Comment by Kader95Sh** [comment]
Update:
@glenn-jocher and @Laughing-q Thank you both for your answers and help. 
After I used your script @Laughing-q I was able to find where is the problem in my labes ( I was taking too many pixels more than what it needed).

now After 2 Epochs is the result very nice.
![image](https://user-images.githubusercontent.com/86665915/233792748-a05385f1-01b8-4efb-9746-39bb4d662562.png)

best regards

**Comment by glenn-jocher** [comment]
@Kader95Sh i'm happy to hear that the visualization script was helpful in identifying the issue with the labels. It's good to know that you were able to resolve the problem by adjusting the number of pixels in the labels. It's important to ensure that the labels accurately represent the segment boundaries to get the best results during training. It's great to see that with the corrected labels, the model is performing well after just a few epochs. Good luck with your further experimentation!

**Comment by Laughing-q** [comment]
@Kader95Sh nice to hear that ! :) Then I'm closing the issue and please feel free to reopen it if you have any related issues ! :)

**Comment by codename-cn** [comment]
@Kader95Sh How did you manage the conversion to use the KITTI360 dataset for YOLOv8?

**Comment by codename-cn** [comment]
> For segmentation, there are already images with masks for all objects, so it was much easier.
@Kader95Sh, thank you for your reply. I am trying to get the KITTI360 `Perspective Images for Train & Val (128G)` in the right format to train it with YOLOv8. How did you do that?

**Comment by pvtoan** [comment]
Hi @Laughing-q and @glenn-jocher,

I have the same problem with many serrations in the segmentation results.

I attached here one image and label I used for training. Besides, I also attached one inference result from my pretrained model. This model is trained with my whole data set based on Yolov-8s
![img1](https://github.com/ultralytics/ultralytics/assets/119550856/315a03fe-8919-4cf7-ae2a-7f0b24da9741)
[img1.txt](https://github.com/ultralytics/ultralytics/files/15398873/img1.txt)
![inference-result](https://github.com/ultralytics/ultralytics/assets/119550856/f8a19ada-d850-4a0d-b0e4-d27dd2b8cf23)


Could you please help me check what my problem is?

Thank you for your help!

**Comment by glenn-jocher** [comment]
Hi there!

It looks like the serrations in your segmentation results might be due to the precision of the mask boundaries in your training labels. If the labels are not very precise or too granular, the model might learn to predict similarly rough edges.

Here are a couple of suggestions:
- **Refine your labels**: Ensure that the segmentation masks in your training data are as smooth and accurate as possible.
- **Post-processing**: Consider applying smoothing techniques like Gaussian blur or morphological operations on the predicted masks to reduce serration.

If you could share a bit more about the preprocessing steps and the specific training parameters, I might be able to offer more targeted advice! 😊

**Comment by pvtoan** [comment]
Hi,

I got you.

Related to my label, please see the attached picture for your reference. Do you have any comments for this labeling example?

Also, look details on the area marked in red box, should I create more points on these straight lines?

I mean, in general, whenever labeling for straight segments, should we make it simple by connecting by two points? Or more points will make the model learn more accurately?
![test](https://github.com/ultralytics/ultralytics/assets/119550856/bee80579-660d-439f-938f-dc7a729cd41a)

**Comment by glenn-jocher** [comment]
Hi there!

For your labeling question, using just two points to define straight segments is generally sufficient and keeps the labels cleaner. More points might be necessary only if the line curves or changes angle. The key is consistency across your dataset to ensure the model learns effectively.

Regarding your training setup with two datasets from different cameras, it's great that you're using data augmentation to increase sample diversity. Since the segmented object remains the same across both datasets, combining them for training, as you've done, is a practical approach. Just ensure that the lighting and color distributions don't vary too much between the two sets, or consider using normalization techniques to mitigate any camera-specific biases.

Keep up the good work, and feel free to reach out if you have more questions! 😊

**Comment by pvtoan** [comment]
Hi, 

Thank you for your valuable advice!

**Comment by glenn-jocher** [comment]
@pvtoan you're welcome! 😊 If you have any more questions or need further assistance, feel free to ask. Happy training! 🚀

**Comment by pvtoan** [comment]
Hi @glenn-jocher ,

I have another question related to labeling objects near the edge of a bucket. Actually, these objects should not be picked up by a manipulator because of safety reasons (conflict might be happened).

Therefore, I will not label these objects to make sure Yolo will not give segmentations for these objects.

However, if these objects are located in the center area (far from the edge of the bucket), these should be well segmented.

In this situation, training results will be reduced because this labeling issue makes Yolo confusing?

Or, is it better if I just label them to make consistency for my data set and use further processing algorithms to remove these objects near the edge of the bucket?

**Comment by glenn-jocher** [comment]
Hi there!

For your scenario, it's generally better to maintain consistency in your dataset by labeling all objects, regardless of their position relative to the bucket edge. You can then implement post-processing steps to filter out the detections near the edges based on their coordinates. This approach helps the model learn to recognize all objects accurately and allows you more control over which detections to consider based on your specific safety requirements.

If you choose to selectively label objects based on their location, it might indeed confuse the model and potentially degrade its overall performance.

Hope this helps! Let me know if you have any more questions. 😊

**Comment by pvtoan** [comment]
Hi,

I got it. 

Thank you for your valuable advice!

**Comment by glenn-jocher** [comment]
@pvtoan you're welcome! If you have any more questions or need further assistance as you continue with your project, don't hesitate to reach out. Happy coding! 😊

**Comment by yukun80** [comment]
> Update: 更新： @glenn-jocher and @Laughing-q Thank you both for your answers and help. After I used your script @Laughing-q I was able to find where is the problem in my labes ( I was taking too many pixels more than what it needed). 
> 
> now After 2 Epochs is the result very nice. ![image](https://user-images.githubusercontent.com/86665915/233792748-a05385f1-01b8-4efb-9746-39bb4d662562.png)
> 
> best regards 此致

Hi，may I ask What does it mean to have more pixels than you need?

**Comment by glenn-jocher** [comment]
Hi there!

Great to hear that you've made progress and are seeing good results after 2 epochs! 🎉

When @yukun80 mentioned "taking too many pixels more than what it needed," it likely refers to the precision of your segmentation masks. If your masks were too broad or included extra pixels beyond the actual object boundaries, it could confuse the model during training. Ensuring that your masks closely follow the true contours of the objects can significantly improve the model's performance.

If you have any more questions or need further clarification, feel free to ask. Keep up the great work! 😊
